### DISCUSSION TITLE
Handling reliable_push buffer full

### DISCUSSION THREAD

**Discussion by stevenharman** [sidekiq/sidekiq#5909] [Ideas]
We are enabling `reliable_push!` and were digging into what exactly happens if we fill up the local, in-memory queue before a connection to Redis can be re-established. Right now, it looks like the job is dropped on the floor (expected) and an `error` level log line is emitted of the form `[ReliablePush] Reached job backup limit, discarding job due to #{ex.class}: #{ex.message}`.

We're wondering if you'd be up for adding instrumentation hooks around the various points of the `Sidekiq::ReliableClient` where we're unable to push a job to Redis. For example, maybe we'd like to log a little more information about the job so we can "re-create" it later, if needed? Or perhaps more likely, allow us to raise an error in the event the job could not be pushed onto the local queue?

**Comment by mperham** [comment]
I'm thinking it would be smarter to raise an error rather than dropping the job. This is the point of backpressure: once the local queue fills up, it should break rather than silently failing. Reliable push is designed for transient network hiccups, not a sustained outage. wdyt?

**Reply by stevenharman** [reply]
Yeah, that was actually what we initially wanted, but it was pretty far away from the current implementation, so decided to ask for something… less. 😄

**Reply by stevenharman** [reply]
I'm imagining something like a new, specific error (perhaps `Sidekiq::ReliableClient::BackupLimitExceeded`?) which will be raised from `ReliableClient#save_locally` when the size is exceeded.

Right now that method logs at `error` level, and returns `false`. Would it still make sense to log _something_ there, and then raise? Or just raise? I could image folks having already built some log-based alerting or something, but 🤷

```ruby
module Sidekiq
  module ReliableClient
    BackupLimitExceeded = Class.new(StandardError)
    
    # … [snip] …
    
    def save_locally(pool, payloads, ex)
      sz = local_queue.size
      if sz > @config[:backup_limit]
        # Currently this just logs and returns false. Instead, log (maybe?) and raise
        @config.logger.error("[ReliablePush] Reached job backup limit, discarding job due to #{ex.class}: #{ex.message}")
        raise BackupLimitExceeded, "Reached job backup limit of (#{sz})"
      else
        # … 
      end
    end

    # … [snip] …

  end
end

**Reply by mperham** [reply]
I like your impl and I think it's a good idea to change the default behavior to `raise`. Returning false swallows the error and makes it quite likely the caller won't see it. Raising gives the system an appropriate amount of backpressure for extended outages. I did add this so people can use metric alerting to find unexpected job discards. I'm not sure it matters though - if Redis is down your entire dashboard is likely bright red already.

```
Sidekiq::Pro.metrics.increment("jobs.discarded.push", by: payloads.size)
```

**Reply by stevenharman** [reply]
Dang... we just hit this case in prod and started swallowing the errors rather than bubbling them up. I think you're on a sabbatical this summer, Mike, but any idea if/when we might be able to get this change in?

**Reply by mperham** [reply]
I'll be back toward the end of the month. It'll ship before Sept.

**Reply by stevenharman** [reply]
Thanks Mike! Enjoy the rest of your summer. 😄

**Reply by mperham** [reply]
Ok, it will now raise `Sidekiq::ReliableClient::BufferFull` if it tries to store too many client payloads. I'm not sure if this is a large enough change in behavior to warrant a version bump to 7.2.0 or if can be part of a patch release. It's an edge case of an edge case.

**Reply by stevenharman** [reply]
This is great Mike. Thank you.

A version bump or patch release… 🤷. Sidekiq is widely-used enough that this is going to break some expectation of someone, somewhere. I guess I'd say point release if this changes documented behavior, or a patch release otherwise.

**Reply by mperham** [reply]
Ok. I'm hoping to ship a patch release next week but I have a few other issues/PRs still pending.
